server/src/main/java/org/elasticsearch/index/mapper/RuntimeField.java

```diff
@@ -58,49 +63,70 @@ default XContentBuilder toXContent(XContentBuilder builder, Params params) throw
+                parameter.toXContent(builder, includeDefaults);
+            }
+            return builder;
         }
 
-        protected abstract RuntimeField createRuntimeField(Mapper.TypeParser.ParserContext parserContext);
-
-        private void validate() {
-            ContentPath contentPath = parentPath(name());
-            FieldMapper.MultiFields multiFields = multiFieldsBuilder.build(this, contentPath);
-            if (multiFields.iterator().hasNext()) {
-                throw new IllegalArgumentException("runtime field [" + name + "] does not support [fields]");
+        public final void parse(String name, Mapper.TypeParser.ParserContext parserContext, Map<String, Object> fieldNode) {
+            Map<String, Parameter<?>> paramsMap = new HashMap<>();
+            for (Parameter<?> param : getParameters()) {
+                paramsMap.put(param.name, param);
             }
-            FieldMapper.CopyTo copyTo = this.copyTo.build();
-            if (copyTo.copyToFields().isEmpty() == false) {
-                throw new IllegalArgumentException("runtime field [" + name + "] does not support [copy_to]");
+            String type = (String) fieldNode.remove("type");
+            for (Iterator<Map.Entry<String, Object>> iterator = fieldNode.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                final String propName = entry.getKey();
+                final Object propNode = entry.getValue();
+                Parameter<?> parameter = paramsMap.get(propName);
+                if (parameter == null) {
+                    if (parserContext.isFromDynamicTemplate() && parserContext.indexVersionCreated().before(Version.V_8_0_0)) {
+                        // The parameter is unknown, but this mapping is from a dynamic template.
+                        // Until 7.x it was possible to use unknown parameters there, so for bwc we need to ignore it
+                        deprecationLogger.deprecate(DeprecationCategory.API, propName,
+                            "Parameter [{}] is used in a dynamic template mapping and has no effect on type [{}]. "
+                                + "Usage will result in an error in future major versions and should be removed.",
+                            propName,
+                            type
+                        );
+                        iterator.remove();
+                        continue;
+                    }
+                    throw new MapperParsingException(
+                        "unknown parameter [" + propName + "] on mapper [" + name + "] of type [" + type + "]"
+                    );
+                }
+                if (propNode == null && parameter.canAcceptNull() == false) {
+                    throw new MapperParsingException("[" + propName + "] on mapper [" + name
+                        + "] of type [" + type + "] must not have a [null] value");
+                }
```

Comment:
I wonder if it's worth factoring out this code that is common between the two implementations instead of duplicating it, though it would mean that the base class has to allow to be extended to introduce parsing of common fields etc.

Reply:
I think the two methods are sufficiently different that it's not worth trying to share stuff here.

Reply:
are they? I did not realize that

Reply:
We have not deprecated anything on the runtime section so far, but I do wonder: should we handle deprecations like we do for field mappers?

Reply:
I guess this makes sense only if we share the code between the two, supporting parameters deprecation while there isn't one deprecated parameter is overkill

Reply:
++ we can copy the machinery from FieldMapper easily enough if we need to